### PR TITLE
Split agent builds & packaging

### PR DESCRIPTION
**Pull request by chouquette**
### What does this PR do?

This PR splits our existing builds in 2 stages:

A build stage, which is common to all agent version X architecture, which produces an unpackaged intermediary artifact
A packaging stage, which produces the expected .deb/.rpm
This is described in more details in [this RFC](https://docs.google.com/document/d/1LksLF6magX4rIf_yDkjW9DDB4oKl8XS8HyNQEyRMkeY/edit?pli=1#heading=h.ixldy9ty2xy5)

### Motivation

Building is long, packaging is cheap, we can save a lot of computing by not building almost identical artifacts for the agent on debian/rhel/suse

### CODE REVIEW

File: release.json [hunk 1]
```diff
@@ -8,7 +8,7 @@
     "nightly": {
         "INTEGRATIONS_CORE_VERSION": "master",
         "OMNIBUS_SOFTWARE_VERSION": "8664357efd9187e1c16284e3d9811ea25d152186",
-        "OMNIBUS_RUBY_VERSION": "78573b961ef3400787a7069526d91795c1a34238",
+        "OMNIBUS_RUBY_VERSION": "chouquette/tmp_split_build_packaging",
```
Comment (kacper-murzyn) [inline release.json:11]: Is this the desired change or something temporary?

**Review by CelianR — APPROVED** [review]

File: release.json (hunk 1 above)
Comment (chouquette) [reply release.json:11]: Temporary, until the required changes in omnibus ruby are merged

**Review by kacper-murzyn — APPROVED** [review]

File: omnibus/config/projects/agent.rb [hunk 2]
```diff
@@ -197,85 +214,79 @@
...
+  # this dependency puts few files out of the omnibus install dir and move them
+  # in the final destination. This way such files will be listed in the packages
+  # manifest and owned by the package manager. This is the only point in the build
+  # process where we operate outside the omnibus install dir, thus the need of
+  # the `extra_package_file` directive.
+  # This must be the last dependency in the project.
+  dependency 'datadog-agent-finalize'
+  dependency 'datadog-cf-finalize'
+  if do_package
+    dependency "init-scripts-agent"
```
Comment (alopezz) [inline omnibus/config/projects/agent.rb:270]: This looks redundant with the conditional right after this (line 273).

File: omnibus/config/projects/agent.rb [hunk 3]
```diff
@@ -197,85 +214,79 @@
...
   extra_package_file '/var/log/datadog/'
 end
 
 # all flavors use the same package scripts
 if linux_target?
+  if do_build && !do_package
+    extra_package_file "#{Omnibus::Config.project_root}/package-scripts/agent-deb"
+    extra_package_file "#{Omnibus::Config.project_root}/package-scripts/agent-rpm"
+  end
   if debian_target?
```
Comment (alopezz) [inline omnibus/config/projects/agent.rb:290]:
```suggestion
  elsif debian_target?
```
I think

File: omnibus/config/software/init-scripts-agent.rb [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+    mkdir "/etc/init"
+    if debian_target?
+      # sysvinit support for debian only for now
+      mkdir "/etc/init.d"
+
+      # debian recommends using a different directory for systemd unit files
+      systemd_directory = "/lib/systemd/system"
+      erb source: "upstart_debian.conf.erb",
+          dest: "/etc/init/datadog-agent.conf",
+          mode: 0644,
```
Comment (alopezz) [inline omnibus/config/software/init-scripts-agent.rb:19]: Just to confirm, we're basically merging two existing steps into one by directly generating the files in their final, location, right?

**Review by alopezz — COMMENTED** [review]
I might still have to make another pass, but if the resulting packages work as expected this is looking good.

**Review by alopezz — COMMENTED** [review]
I might still have to make another pass, but if the resulting packages work as expected this is looking good.

File: omnibus/config/projects/agent.rb (hunk 2 above)
Comment (chouquette) [reply omnibus/config/projects/agent.rb:270]: Ah indeed. It's not entirely redundant as we need the one on line 273 for heroku (which for now still does build & packaging in a single step) but the flow is incorrect as we shouldn't care about packaging if we were building.

I'll push a fix shortly

File: omnibus/config/projects/agent.rb (hunk 3 above)
Comment (chouquette) [reply omnibus/config/projects/agent.rb]: I enclosed the 2 distribution specific lines in an `if do_package` block, as they don't make sense when building

**Review by alopezz — APPROVED** [review]

**Review by paulcacheux — APPROVED** [review]

**Pull request merged**